### DISCUSSION TITLE
How to Run DAG Steps in Parallel While Passing Data In-Memory in Airflow?

### DISCUSSION THREAD

**Discussion by takerfume** [apache/airflow#35296] [General] [answered]
I'm new to Airflow and am trying to understand its capabilities. I recently asked [a similar question](https://github.com/dagster-io/dagster/discussions/17494) in the Dagster community regarding the execution of DAG steps in parallel while also passing data between the steps in-memory, rather than reading and writing from the disk. This is crucial for my use-case (processing many images), as both avoiding I/O operations and achieving parallel execution are important to prevent bottlenecks.

Is it possible to achieve these objectives in Airflow? Specifically, can multiple tasks in a DAG be executed in parallel while passing data between them in-memory?
I would like to use python.

**Comment by Taragolis** [comment, marked as answer]
There is no buitin in-memory access and this unlikely it would be implemented in foreseeable future. Airflow tasks isolated from each other and depend on executor could be executed in different machines as result all communications between the task are done through [XCom](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/xcoms.html) which is no suits to your requirements.

In theory you could implement custom [XCom Backend](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/xcoms.html#custom-xcom-backends) and use LocalExecutor

**Reply by takerfume** [reply]
Thank you for your response!

I will try dagseter's custom io manager with redis which maintainer recommended to me in the following comment.
https://github.com/dagster-io/dagster/discussions/17494#discussioncomment-7439572
